### PR TITLE
Allow Wire to switch from master to slave and vice versa

### PR DESCRIPTION
**Pull request by maidnl**
This PR solve https://github.com/arduino/ArduinoCore-renesas/issues/180 and supersedes https://github.com/arduino/ArduinoCore-renesas/pull/182.
Basically allow I2C to switch from master to slave and vice versa in a cleaner way (with less changes).

### CODE REVIEW

**Review by aentinger — APPROVED** [review]
LGTM :+1: 

So clean ... :heart_eyes:

File: libraries/Wire/Wire.cpp [hunk 1]
```diff
@@ -404,26 +410,23 @@ void TwoWire::begin(void) {
...
   /* Address is set inside begin() using slave_address member variable */
-  begin();
+  _begin();
   
 }
 
 /* -------------------------------------------------------------------------- */
 void TwoWire::begin(int address) {
 /* -------------------------------------------------------------------------- */  
-  is_master = false;
```
Comment (facchinm) [inline libraries/Wire/Wire.cpp:417]: Is this change necessary? What happens in case the caller requests slave -> master -> slave ?

File: libraries/Wire/Wire.cpp [hunk 2]
```diff
@@ -273,6 +273,16 @@ bool TwoWire::cfg_pins(int max_index) {
 
 /* -------------------------------------------------------------------------- */
 void TwoWire::begin(void) {
+/* -------------------------------------------------------------------------- */  
+  end();
```
Comment (facchinm) [inline libraries/Wire/Wire.cpp:277]: Is end() needed here? Should we also add it to `begin(address)` ?

**Review by facchinm — CHANGES_REQUESTED** [review]

File: libraries/Wire/Wire.cpp (hunk 1 above)
Comment (maidnl) [reply libraries/Wire/Wire.cpp:417]: This is a simplification (I removed duplicated code). Before there were 3 different `begin(address)` functions and all of them were setting is_master = false, but in the end the all finished up to call the version `begin(uint16_t address)`. So I removed 2 lines of code from the other 2 two (master = false and slave_address = address). You can change role as many time as you want.

File: libraries/Wire/Wire.cpp (hunk 2 above)
Comment (maidnl) [reply libraries/Wire/Wire.cpp:277]: end() is called in both cases -> in case of master initialization in the line you spotted, in case of slave initialization is called only in the version of begin(uint16_t address) (line 413), because the other 2 begin functions (i.e. begin(int address) and begin(uint8_t address) are in the end calling the version with uint16_t as argument.

**Review by facchinm — APPROVED** [review]
